renderdoc/driver/ihv/intel/intel_counters.cpp

```diff
@@ -0,0 +1,437 @@
+/******************************************************************************
+* The MIT License (MIT)
+*
+* Copyright (c) 2018 Baldur Karlsson
+*
+* Permission is hereby granted, free of charge, to any person obtaining a copy
+* of this software and associated documentation files (the "Software"), to deal
+* in the Software without restriction, including without limitation the rights
+* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+* copies of the Software, and to permit persons to whom the Software is
+* furnished to do so, subject to the following conditions:
+*
+* The above copyright notice and this permission notice shall be included in
+* all copies or substantial portions of the Software.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+* THE SOFTWARE.
+******************************************************************************/
+
+#include "common/common.h"
+
+#include "official/DriverStorePath.h"
+#include "intel_counters.h"
+
+#include <set>
+
+const static std::vector<std::string> metricSetBlacklist = {"TestOa"};
+
+HMODULE IntelCounters::m_MDLibraryHandle;
+IMetricsDevice_1_5 *IntelCounters::m_metricsDevice;
+OpenMetricsDevice_fn IntelCounters::OpenMetricsDevice;
+CloseMetricsDevice_fn IntelCounters::CloseMetricsDevice;
+
+IntelCounters::IntelCounters()
+    : m_device(NULL), m_deviceContext(NULL), m_counter(NULL), m_passIndex(0), m_sampleIndex(0)
+{
+}
+
+IntelCounters::~IntelCounters()
+{
+  if(CloseMetricsDevice)
+    CloseMetricsDevice(m_metricsDevice);
+
```

Comment:
`m_metricsDevice` isn't set to `NULL` here when it's closed, so I think next time around `IntelCounters::Load` won't open it again and we'll try to use an invalid pointer.

Reply:
Added metrics device initialization/invalidation.